crates/red_knot_python_semantic/resources/mdtest/statically-known-branches.md

```diff
@@ -0,0 +1,684 @@
+    else:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+```py path=nested_if_true_if_false.py
+x = 1
+
+if True:
+    if False:
+        x = 2
+    else:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[3]
+```
+
+```py path=nested_if_true_if_bool.py
+def flag() -> bool: ...
+
+x = 1
+
+if True:
+    if flag():
+        x = 2
+    else:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[2, 3]
+```
+
+```py path=nested_if_bool_if_true.py
+def flag() -> bool: ...
+
+x = 1
+
+if flag():
+    if True:
+        x = 2
+    else:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[2, 4]
+```
+
+```py path=nested_else_if_true.py
+x = 1
+
+if False:
+    x = 2
+else:
+    if True:
+        x = 3
+    else:
+        x = 4
+
+reveal_type(x)  # revealed: Literal[3]
+```
+
+```py path=nested_else_if_false.py
+x = 1
+
+if False:
+    x = 2
+else:
+    if False:
+        x = 3
+    else:
+        x = 4
+
+reveal_type(x)  # revealed: Literal[4]
+```
+
+```py path=nested_else_if_bool.py
+def flag() -> bool: ...
+
+x = 1
+
+if False:
+    x = 2
+else:
+    if flag():
+        x = 3
+    else:
+        x = 4
+
+reveal_type(x)  # revealed: Literal[3, 4]
+```
+
+### Combination with non-conditional control flow
+
+#### `try` ... `except`
+
+```py path=try_if_true.py
+def may_raise() -> None: ...
+
+x = 1
+
+try:
+    may_raise()
+    if True:
+        x = 2
+    else:
+        x = 3
+except:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[2, 4]
+```
+
+```py path=if_true_try.py
+def may_raise() -> None: ...
+
+x = 1
+
+if True:
+    try:
+        may_raise()
+        x = 2
+    except:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[2, 3]
+```
+
+#### `for` loops
+
+```py path=for_if_true.py
+def iterable() -> list[()]: ...
+
+x = 1
+
+for _ in iterable():
+    if True:
+        x = 2
+    else:
+        x = 3
+
+reveal_type(x)  # revealed: Literal[1, 2]
+```
+
+```py path=for_else_if_true.py
+def iterable() -> list[()]: ...
+
+x = 1
+
+for _ in iterable():
+    x = 2
+else:
+    if True:
+        x = 3
+    else:
+        x = 4
+
+reveal_type(x)  # revealed: Literal[3]
+```
+
+```py path=if_true_for.py
+def iterable() -> list[()]: ...
+
+x = 1
+
+if True:
+    for _ in iterable():
+        x = 2
+else:
+    x = 3
+
+reveal_type(x)  # revealed: Literal[1, 2]
+```
+
+```py path=if_true_for_else.py
+def iterable() -> list[()]: ...
+
+x = 1
+
+if True:
+    for _ in iterable():
+        x = 2
+    else:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[3]
+```
+
+## If expressions
+
+See also: tests in [expression/if.md](expression/if.md).
+
+### Always true
+
+```py
+x = 1 if True else 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+### Always false
+
+```py
+x = 1 if False else 2
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+## Boolean expressions
+
+### Always true, `or`
+
+```py
+(x := 1) == 1 or (x := 2)
```

Comment:
This should have the same behavior without the `== 1`, right? Worth testing?

Reply:
Correct. Changed to the more concise form.